libs/providers/flagd/src/lib/service/grpc/service.ts

```diff
@@ -2,111 +2,238 @@ import * as grpc from '@grpc/grpc-js';
-      flagKey,
-      context: this.convertContext(context, logger),
-    }).then(this.onFulfilled, this.onRejected);
-    return {
-      value: response.value,
-      reason: response.reason,
-      variant: response.variant,
-    }
+  async resolveString(flagKey: string, context: EvaluationContext, logger: Logger): Promise<ResolutionDetails<string>> {
+    return this.resolve(this._client.resolveString, flagKey, context, logger);
   }
 
-  async resolveNumber(
-    flagKey: string,
-    context: EvaluationContext,
-    logger: Logger
-  ): Promise<ResolutionDetails<number>> {
-    const { response } = await this.client.resolveFloat({
-      flagKey,
-      context: this.convertContext(context, logger),
-    }).then(this.onFulfilled, this.onRejected);
-    return {
-      value: response.value,
-      reason: response.reason,
-      variant: response.variant,
-    };
+  async resolveNumber(flagKey: string, context: EvaluationContext, logger: Logger): Promise<ResolutionDetails<number>> {
+    return this.resolve(this._client.resolveFloat, flagKey, context, logger);
   }
 
   async resolveObject<T extends JsonValue>(
     flagKey: string,
     context: EvaluationContext,
     logger: Logger
   ): Promise<ResolutionDetails<T>> {
-    const { response } = await this.client.resolveObject({
-      flagKey,
-      context: this.convertContext(context, logger),
-    }).then(this.onFulfilled, this.onRejected);
-    if (response.value !== undefined) {
-      return {
-        value: Struct.toJson(response.value) as T,
-        reason: response.reason,
-        variant: response.variant,
-      };
+    return this.resolve(this._client.resolveObject, flagKey, context, logger, this.objectParser);
+  }
+
+  private connectStream(): Promise<boolean> {
+    return new Promise((resolve, reject) => {
+      this.logger?.info(`${FlagdProvider.name}: connecting stream, attempt ${this._streamConnectAttempt}...`);
+      const stream = this._client.eventStream({});
+      stream.responses.onError(() => {
+        this.handleError(reject);
+      });
+      stream.responses.onComplete(() => {
+        this.handleComplete();
+      });
+      stream.responses.onMessage((message) => {
+        if (message.type === EVENT_PROVIDER_READY) {
+          this.handleProviderReady(resolve);
+        } else if (message.type === EVENT_CONFIGURATION_CHANGE) {
+          this.handleFlagsChanged(message);
+        }
+      })
+    });
+  }
+
+  private handleProviderReady(resolve: (value: boolean | PromiseLike<boolean>) => void) {
+    this.logger?.info(`${FlagdProvider.name}: streaming connection established with flagd`);
+    this._streamAlive = true;
+    this._streamConnectAttempt = 0;
+    this._streamConnectBackoff = BASE_EVENT_STREAM_RETRY_BACKOFF_MS;
+    resolve(true);
+  }
+
+  private handleFlagsChanged(message: EventStreamResponse) {
+    if (message.data) {
+      const parsedMessage = Struct.toJson(message.data) as FlagChangeMessage;
```

Comment:
Is this a safe case? Should there be a type guard?

Reply:
We have all the null checks we need later, but this does at least _look_ suspect. I will improve it.

Reply:
fixed.